crates/iota-json-rpc/src/coin_api.rs

```diff
@@ -383,6 +376,70 @@ async fn find_package_object_id(
     .await?
 }
 
+#[instrument(skip(internal), ret)]
+async fn gas_total_supply<I>(internal: &I, tag: &StructTag) -> Result<Option<Supply>, Error>
+where
+    I: CoinReadInternal + Send + Sync + ?Sized,
+{
+    if !GAS::is_gas(tag) {
+        return Ok(None);
+    }
+
+    let summary = IotaSystemStateSummaryV2::try_from(
+        internal
+            .get_state()
+            .get_system_state()
+            .map_err(Error::from)?
+            .into_iota_system_state_summary(),
+    )
+    .map_err(Error::from)?;
+
+    Ok(Some(Supply {
+        value: summary.iota_total_supply,
+    }))
+}
+
+#[instrument(skip(internal), ret)]
+async fn treasury_cap_total_supply<I>(
+    internal: &I,
+    tag: &StructTag,
+) -> Result<Option<Supply>, Error>
+where
+    I: CoinReadInternal + Send + Sync + ?Sized,
+{
+    if let Ok(obj) = internal
+        .find_package_object(&tag.address.into(), TreasuryCap::type_(tag.clone()))
+        .await
+    {
+        let data = obj
+            .data
+            .try_as_move()
+            .ok_or_else(|| Error::Unexpected("Cannot get move contents".into()))?
+            .contents();
+        let tc = TreasuryCap::from_bcs_bytes(data).map_err(Error::from)?;
```

Comment:
Similarly to `CoinManager` we could implement `TryFrom<Object>` and consolidate `coin_manager_total_supply` and `treasury_cap_total_supply` to use a common method for both types generic on a type `T: TryFrom<Object>`